src/hotspot/os/posix/signals_posix.cpp

```diff
@@ -1606,11 +1583,11 @@ static void SR_handler(int sig, siginfo_t* siginfo, ucontext_t* context) {
 
   // On some systems we have seen signal delivery get "stuck" until the signal
   // mask is changed as part of thread termination. Check that the current thread
-  // has not already terminated (via SR_lock()) - else the following assertion
+  // has not already terminated - else the following assertion
   // will fail because the thread is no longer a JavaThread as the ~JavaThread
   // destructor has completed.
 
-  if (thread->SR_lock() == NULL) {
+  if (thread->is_Java_thread() && thread->as_Java_thread()->is_terminated()) {
     return;
   }
```

Comment:
If JavaThread destructor have been run the JavaThread pointer is invalid, so we may not dereference it.
Thus the old code is invalid if SR_lock is NULL, because you just dereferenced a pointer to a destroyed object.

If the user of OS Suspend/Resume (e.g. JFR) have it on a ThreadsList it cannot become terminated.
If it's terminated it is not on any newly created ThreadsList, it cannot be seem, anyways.
Assuming the user uses a ThreadsList the code is valid, a user without a ThreadsList this code is still dereferencing a pointer to a destroyed object, but there is no easy fix here for this.
If the user promise to use a ThreadsList we can remove this check.

So instead of trying to make something equivalent I tried to make it 'better' instead.

Sorry I don't see why the assert won't fire?
We used to bailout for any Thread, now we only bailout for JavaTread (if terminated), so the scope have increased.
We now fire always for a non Java/VM thread, which we did not do before.